File: ui/src/routes/api/agents.ts
import { defineApi } from '@/server/api-route'
import { json } from '@/server/http'
import { listFleetAgents } from '@/server/fleet-agents'
import { requireUser } from '@/server/api-guard'
import { usableAgentGate } from '@/server/users'

// GET /api/agents → the fleet the current user may use (definition-backed
// agents with their model tiers, filtered by per-agent access). Auth-gated.
export const Route = defineApi('/api/agents', {
  GET: async ({ request }) => {
    const user = await requireUser(request)
    if (user instanceof Response) return user

    const agents = await listFleetAgents()
    // Owner-aware: a personal assistant is only visible to its owner.
    const usable = await usableAgentGate(user.id, user.role)
    const visible = agents.filter((a) => usable(a.id))
    return json({ agents: visible })
  },
})

File: api/src/routes/agents/agents.rs
// GET /api/agents. The fleet the current user may use (definition-backed
// agents with their model tiers, filtered by per-agent access). Auth-gated.

use crate::error::thrown_internal_error;
use crate::fleet::{list_fleet_agents, usable_agent_gate};
use crate::session::require_user;
use crate::state::AppState;
use axum::Json;
use axum::extract::State;
use axum::http::HeaderMap;
use axum::response::{IntoResponse, Response};

#[derive(serde::Serialize)]
struct AgentsBody {
    agents: Vec<crate::fleet::FleetAgentEntry>,
}

pub async fn get(State(state): State<AppState>, headers: HeaderMap) -> Response {
    let user = match require_user(&state, &headers).await {
        Ok(u) => u,
        Err(gate) => return gate,
    };
    // Owner-aware: a personal assistant is only visible to its owner.
    let (agents, gate) = tokio::join!(
        list_fleet_agents(&state.pg),
        usable_agent_gate(&state.pg, &user.id, &user.role),
    );
    let (agents, gate) = match (agents, gate) {
        (Ok(a), Ok(g)) => (a, g),
        (Err(e), _) | (_, Err(e)) => {
            tracing::error!("[agents] fleet read failed: {e}");
            return thrown_internal_error();
        }
    };
    let visible = agents.into_iter().filter(|a| gate(&a.agent.id)).collect();
    Json(AgentsBody { agents: visible }).into_response()
}
